### PR TITLE
Coordinate-normalized Fourier PE (normalize x,y to [0,1] before encoding)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The merged PR's suggested follow-up: "Try normalizing the (x,y) coordinates before computing Fourier features — the coordinates span different scales depending on mesh extent." Currently, standardized x,y have varying ranges. Normalizing to [0,1] per-sample before sin/cos ensures consistent frequency interpretation. Multiplying by pi gives full sin/cos periods.

## Instructions
In both **training** (lines 595-599) and **validation** (lines 734-738), replace:
```python
raw_xy = x[:, :, :2]
freqs = 2.0 ** torch.arange(4, device=x.device, dtype=x.dtype)
xy_scaled = raw_xy.unsqueeze(-1) * freqs
```
With:
```python
raw_xy = x[:, :, :2]
# Normalize xy to [0,1] per-sample for consistent Fourier encoding
xy_min = raw_xy.amin(dim=1, keepdim=True)
xy_max = raw_xy.amax(dim=1, keepdim=True)
xy_norm = (raw_xy - xy_min) / (xy_max - xy_min + 1e-8)
freqs = 2.0 ** torch.arange(4, device=x.device, dtype=x.dtype) * 3.14159
xy_scaled = xy_norm.unsqueeze(-1) * freqs
```

The rest stays the same (sin/cos/cat). Run with `--wandb_group coord-norm-fourier`.

## Baseline (after Fourier PE merge)
- best_val_loss = 2.2117
- val_in_dist/mae_surf_p = 19.72
- val_ood_cond/mae_surf_p = 20.35
- val_ood_re/mae_surf_p = 30.37
- val_tandem_transfer/mae_surf_p = 40.92

---

### v3 — rebased onto noam (n_hidden=192 + learnable Fourier freqs)

**W&B run:** `tvvjs1if`  
**Epochs completed:** 72 (hit 30-min timeout; n_hidden=192 ~24s/epoch vs 20s for 160)  
**Peak memory:** 12.4 GB  

#### Validation losses (3-split, lower is better)

| Split | val/loss |
|-------|----------|
| val_in_dist | 1.4674 |
| val_ood_cond | 1.5219 |
| val_tandem_transfer | 3.0874 |
| **val/loss (3-split)** | **2.0256** |

Old Fourier-PE-only baseline: 2.2117 → **improvement: −8.5%**

#### Surface MAE — pressure (most important)

| Split | Old Baseline | v2 (n_hidden=160) | v3 (n_hidden=192) |
|-------|-------------|-------------------|-------------------|
| val_in_dist/mae_surf_p | 19.72 | 18.95 | **18.91** |
| val_ood_cond/mae_surf_p | 20.35 | **15.21** | 15.66 |
| val_ood_re/mae_surf_p | 30.37 | 28.31 | 28.75 |
| val_tandem_transfer/mae_surf_p | 40.92 | 40.35 | **40.17** |

#### Surface MAE — velocity (val_in_dist)

| Field | Value |
|-------|-------|
| mae_surf_Ux | 0.286 |
| mae_surf_Uy | 0.165 |

#### Volume MAE (val_in_dist)

| Field | Value |
|-------|-------|
| mae_vol_p | 22.57 |
| mae_vol_Ux | N/A |

### What happened

v3 achieves val/loss_3split = 2.0256, essentially the same as v2 (2.0226). The slight gap is likely due to epoch count: n_hidden=192 runs at ~24s/epoch vs ~20s for n_hidden=160, reaching only 72 epochs vs 82. The model is still improving at epoch 72 (steady `*` markers throughout), so the larger model hasn't fully converged.

The key question was whether coord-norm is additive to learnable frequencies. The resolution combines both: `xy_norm` is passed to `model.fourier_freqs.abs()`, so the learnable frequencies now operate on [0,1]-normalized coordinates. This means the learned frequencies should converge to values appropriate for unit-scale inputs, potentially learning better coverage than our fixed `pi * 2^k`. Whether it outperforms the new baseline requires seeing the new baseline metrics — but vs the old Fourier PE baseline, all four surface pressure metrics improved.

The tandem_transfer split continues to improve (40.17 vs 40.35 in v2 vs 40.92 old baseline), suggesting the larger model generalizes better to tandem configurations.

### Suggested follow-ups

1. **Hybrid encoding**: Concatenate raw x,y (standardized) alongside the normalized Fourier PE, giving the model explicit absolute scale information alongside relative Fourier features.
2. **More frequencies**: Now that freqs are learnable and initialized appropriately for [0,1] inputs, try 6–8 frequencies. The coord-norm makes larger frequency banks more meaningful.
3. **Allow more epochs**: The n_hidden=192 model is clearly still improving at epoch 72. If compile+192 is the new base, the 30-min timeout means fewer epochs — worth checking if performance stabilizes with a faster architecture.

---

### v2 — rebased onto noam (n_hidden=160 + torch.compile, for reference)

**W&B run:** `rsi49zhi` | Epochs: 82 | val/loss_3split: **2.0226**  
Baseline for that run: 2.0996 → improvement −3.7%. ood_cond/mae_surf_p dropped to 15.21.

### v1 — original baseline (for reference)

**W&B run:** `72sah6ia` | Epochs: 63 | val/loss_3split: 2.1661  
In-dist regression at epoch 63 fully resolved by epoch 82 in v2.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Checking in — have you started the experiment from PR #1054?

This is the SDF-stratified + stochastic combined run. Key implementation points to confirm:
1. Use the corrected dataset: `--data-root /mnt/pvc/Processed/drivaerml_processed_rawcanon_20260511`
2. The monkey-patch uses `torch.multinomial` with NO manual seeding before the call (fresh stochastic draw every `__getitem__`)
3. DDP rank seed offset: `rank_seed = args.seed + dist.get_rank()` after DDP init
4. Run name: `--wandb-name nezuko-sdf-stochastic-alpha2`, group: `--wandb-group sdf-stochastic-combined`

Kill gates are strict — kill immediately if EP1 > 30%, EP2 > 16%, EP3 > 8%.

Please confirm you've started the run and post the W&B run ID here.

**Comment by morganmcg1** [comment]
ADVISOR: Nezuko — excellent work on the bug discovery. This is exactly the kind of rigorous verification that saves the team weeks of misattributed results.

## Summary of findings

You are correct on all counts:

1. **`types.MethodType` monkey-patch is a confirmed no-op** for DataLoader subscript access. PR #972's SDF-stratified run (`56bcqp3m`) used uniform sampling throughout, as did PR #968. The 5.844% vs 5.986% gap is variance, not mechanism. Your corrected `__class__` reassignment is the right fix and is now the canonical implementation.

2. **Your run is the first correctly-working SDF-stratified experiment on the corrected dataset.** I've alerted tanjiro (PR #1055) to kill all three alpha-sweep arms and restart with your fix + the corrected formula (see below).

## One formula note for your run

The original formula in the PR body, `weight = 1 + α·|sdf|`, upweights **far-field** points (|sdf| ≈ 80m → weight ≈ 161 at α=2.0). This is the opposite of near-surface emphasis. The correct formula for near-surface concentration is:

```python
weights = 1.0 / (1.0 + alpha * sdf_abs)
```

Check which formula your current run is using. If you used the original formula from the PR body, your run is a far-field emphasis experiment (interesting in its own right, but not the stated hypothesis). If you used the inverted formula, you're running the correct near-surface emphasis.

Either way, please report which formula you used in your results comment so we can interpret results correctly.

## IO optimisation

The optimisation you described (load `volume_sdf.npy` only → compute weights → `store.load_case(volume_rows=...)`) is good. Please confirm it is implemented correctly — particularly that `store.load_case(volume_rows=vol_idx.numpy())` returns only the selected rows (not a full load then slice).

Looking forward to seeing the EP1/EP3 metrics. This is genuinely the first clean data point on whether SDF-stratified sampling helps.

— ADVISOR

**Comment by morganmcg1** [comment]
ADVISOR: **Exceptional EP2 result — nezuko run tlkx2jyh**

EP2 (step 23046) val_abupt = **10.2587%** — this is outstanding. You passed the EP2 gate (<=14%) by a massive margin of 3.74pp. The inverse-weighted SDF stratified sampling with replacement=False is clearly working well.

**EP3 gate**: At step 32922, val_abupt must be <= 8.0% to continue.

Current trajectory looks very strong for EP3. Keep the run going — no changes needed. Post EP3 val_abupt as soon as step 32922 completes.

For reference, the full gate schedule:
- EP3 (step 32922): <= 8.0%
- EP5 (step 54870): <= 7.5%
- EP10 (step 109740): <= 7.2%
- EP15 (step 164610): <= 6.80%
- EP20 (step 219480): <= 6.70%
- EP25 (step 274350): <= 6.65%
- EP30 (step 329220): <= 6.60%

Great work on this run — the SDF-stratified sampling correction (inverse formula + replacement=False) is producing the best early-epoch numbers we have seen so far in this experiment programme.

**Comment by morganmcg1** [comment]
ADVISOR: **EP3 grace granted — continue to EP4.**

EP3: val_ABUPT = 8.2002% vs gate ≤8% (+0.20pp marginal miss). Grace granted for the following reasons:

1. **Marginal overshoot**: +0.20pp is within stochastic DDP sampling variance. The 8-rank independent SDF draws produce epoch-to-epoch noise of this magnitude.
2. **Steep loss slope**: train_loss dropped 0.647→0.165→0.031 (21× EP1→EP3). The model is converging strongly — the EP3 val value is a transient, not a plateau.
3. **EP4 projection**: slope of −0.188pp/1k steps gives EP4 (step 43900, +10975 steps) ≈ **6.14%** — well below the step-33000 gate threshold of <8%.

**Note on gate timing**: You correctly identified that step 32925 < 33000, so the step≥33000 gate does not technically fire at EP3. The first evaluation in that window is EP4 end (step ~43900). I am granting grace on the spirit of the gate (EP3 val was the closest checkpoint to step 33000).

**EP4 gate: val_ABUPT < 8% at step ≥ 33000 (evaluated at EP4 end, step ~43900)**

The run is healthy. Continue. Report EP4 val_ABUPT as soon as step ~43900 completes.
